portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java

```diff
@@ -0,0 +1,127 @@
+
+  /** The name param is string property identifier for setting and retrieving
+   * the name of the user associated with a submitted comment */
+  private static final String NAME_PARAM = "name";
+
+  /** The time param is a string property identifier for the time of submission 
+   * upon a POST request for a submitted comment. It is also the property used
+   * to sort comments in most recently submitted order */
+  private static final String TIME_PARAM = "timestamp";
+  
+  /** The tag param is a string property identifier for comments. It is the 
+   * property used to determine which comments are displayed with their 
+   * corresponding blog posts. If not included in a GET request, no comments
+   * will be returned. */
+  private static final String TAG_PARAM = "tag";
+
+  /** The url param is a string identifier for the page url that the associated 
+   * comments will be posted to and where the user will be redirected to after 
+   * posting a comment. If not included POST request, user will not be 
+   * redirected to the original page after submitting a comment */
+  private static final String URL_PARAM = "current-url";
+
+  /** The comment param specifices the string identifier for the entity class 
+   * that the GET request queries */
+  private static final String COMMENT = "Comment";
+
+  /** The path len param specifices the integer identifier for the length of
+   * the blog url path up until its tag specifier. If not included in the POST
+   * request, the blog post tag cannot be retrieved for subsequent association
+   * with the submitted comment */
+  private static final int PATH_LEN = 15;
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String requestTag = request.getParameter(TAG_PARAM);
+    Query.FilterPredicate filter = new Query.FilterPredicate(TAG_PARAM, FilterOperator.EQUAL, requestTag);
+    Query query = new Query(COMMENT).setFilter(filter).addSort(TIME_PARAM, SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    
+    List<Comment> comments = new ArrayList<>();
+
+    if (requestTag != null) {
+      for (Entity entity : results.asIterable()) {
+        String tag = (String) entity.getProperty(TAG_PARAM);
+        String name = (String) entity.getProperty(NAME_PARAM);
+        String text = (String) entity.getProperty(TEXT_PARAM);
+        long timestamp = (long) entity.getProperty(TIME_PARAM);
+        Comment comment = new Comment(name, text, tag, timestamp);
+        comments.add(comment);
+      }
+    }
+
+    String json = convertToJson(comments);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String text = request.getParameter(TEXTIN_PARAM);
```

Comment:
We've probably mentioned in the past, but some validation may be nice. Even if it's not HTML escaping, just checking that these aren't null or empty string would save you from having invalid entries in your database.